File: app/src/test/java/com/dpis/module/FontLibraryActivitySourceSmokeTest.java
package com.dpis.module;

import org.junit.Test;

import java.io.IOException;

import static org.junit.Assert.assertTrue;
import static org.junit.Assert.assertFalse;

public final class FontLibraryActivitySourceSmokeTest {
    @Test
    public void ttcImportIsStandardAndRegistersAllLoadableFaces() throws IOException {
        String source = read("src/main/java/com/dpis/module/fonts/FontLibraryActivity.java");
        String strings = read("src/main/res/values/strings.xml");

        assertTrue(source.contains("lowerName.endsWith(\".ttc\")"));
        assertTrue(source.contains("\"font/ttc\".equals(mimeType)"));
        assertTrue(source.contains("FontFileInspector.inspect(tempFile)"));
        assertTrue(source.contains("findLoadableTtcFaceIndexes("));
        assertTrue(source.contains("fontLibraryStore.registerCopiedFontFaces("));
        assertTrue(source.contains("FontTypefaceLoader.load(file, index)"));
        assertTrue(source.contains("font_library_import_count_success"));
        assertFalse(source.contains("isTtcFontImportEnabled"));
        assertFalse(source.contains("showTtcFaceSelectionDialog"));
        assertFalse(strings.contains("font_library_ttc_select_title"));
    }

    @Test
    public void fontImportNameUsesLargeDialogWidth() throws IOException {
        String source = read("src/main/java/com/dpis/module/fonts/FontLibraryActivity.java");

        assertTrue(source.contains("private void promptImportName("));
        assertTrue(source.contains("ComposeTextInputDialog.showLarge(this"));
    }

    @Test
    public void oversizedFontUsesConfirmationInsteadOfHardRejection() throws IOException {
        String source = read("src/main/java/com/dpis/module/fonts/FontLibraryActivity.java");
        String strings = read("src/main/res/values/strings.xml");

        assertTrue(source.contains("LARGE_FONT_WARNING_BYTES"));
        assertTrue(source.contains("confirmLargeFontImport("));
        assertTrue(source.contains("resolveDocumentSize(uri)"));
        assertTrue(source.contains("hasImportSpace(sizeBytes)"));
        assertTrue(source.contains("sourceSizeBytes * 3L + IMPORT_FREE_SPACE_MARGIN_BYTES"));
        assertTrue(source.contains("font_library_large_import_continue"));
        assertTrue(strings.contains("font_library_large_import_message"));
        assertTrue(strings.contains("font_library_import_insufficient_space"));
    }

    @Test
    public void fontDetailsDisplayPublicationStatus() throws IOException {
        String source = read("src/main/java/com/dpis/module/fonts/FontDetailActivity.java");
        String content = read("src/main/java/com/dpis/module/ui/compose/FontLibraryContent.kt");
        String strings = read("src/main/res/values/strings.xml");

        assertTrue(source.contains("new FontDetailUiState("));
        assertTrue(source.contains("entry.sourceFileName == null ? \"\" : entry.sourceFileName"));
        assertTrue(strings.contains("font_library_private_badge"));
        assertTrue(strings.contains("font_library_public_badge"));
        assertTrue(content.contains("state.publicationFailed"));
        assertTrue(content.contains("font_library_private_badge"));
        assertTrue(content.contains("font_library_public_badge"));
        assertTrue(source.contains("showFallbackExplanationDialog"));
        assertTrue(source.contains("FontPublicationStatus.PUBLISH_FAILED"));
        assertTrue(source.contains("retryPublishedFallbacks"));
        assertTrue(strings.contains("font_library_publication_retry_action"));
        assertTrue(strings.contains("font_library_fallback_dialog_message"));
        assertFalse(source.contains("resolveFontSubtitle(entry)"));
    }

    @Test
    public void fontLibraryGroupsFacesIntoCollections() throws IOException {
        String source = read("src/main/java/com/dpis/module/fonts/FontLibraryActivity.java");

        assertTrue(source.contains("LinkedHashMap<String, List<FontLibraryEntry>> collections"));
        assertTrue(source.contains("new FontLibraryUiItem("));
        assertTrue(source.contains("faces.size() > 1"));
        assertTrue(source.contains("font_library_collection_label"));
    }

    @Test
    public void fontLibraryImportsAndExportsSeparateArchives() throws IOException {
        String source = read("src/main/java/com/dpis/module/fonts/FontLibraryActivity.java");
        String content = read("src/main/java/com/dpis/module/ui/compose/FontLibraryContent.kt");

        assertTrue(content.contains("ic_more_vert_24"));
        assertTrue(content.contains("font_library_export_archive_action"));
        assertTrue(content.contains("font_library_import_archive_action"));
        assertTrue(source.contains("openFontLibraryExportPicker"));
        assertTrue(source.contains("openFontLibraryImportPicker"));
        assertTrue(source.contains("FontLibraryArchiveCodec.writeArchive"));
        assertTrue(source.contains("FontLibraryArchiveCodec.restoreArchive"));
        assertTrue(source.contains("REQUEST_EXPORT_FONT_LIBRARY"));
        assertTrue(source.contains("REQUEST_IMPORT_FONT_LIBRARY"));
        assertTrue(source.contains("recoverMissingFontCatalogAsync"));
        assertTrue(source.contains("recoverMissingCatalogEntries"));
    }

    @Test
    public void fontDetailsUseDedicatedActivityAndFullPageLayout() throws IOException {
        String library = read("src/main/java/com/dpis/module/fonts/FontLibraryActivity.java");
        String detail = read("src/main/java/com/dpis/module/fonts/FontDetailActivity.java");
        String content = read("src/main/java/com/dpis/module/ui/compose/FontLibraryContent.kt");
        String manifest = read("src/main/AndroidManifest.xml");

        assertTrue(library.contains("new Intent(this, FontDetailActivity.class)"));
        assertTrue(library.contains("FontDetailActivity.EXTRA_FONT_ID"));
        assertTrue(!library.contains("showFontDetails("));
        assertTrue(detail.contains("EXTRA_FONT_ID"));
        assertTrue(detail.contains("showFallbackExplanationDialog"));
        assertTrue(detail.contains("confirmDeleteForCurrentEntry"));
        assertTrue(detail.contains("confirmClearAppTypefaceByPackage"));
        assertTrue(content.contains("FontReferenceSection("));
        assertTrue(content.contains("font_library_active_apps_title"));
        assertTrue(content.contains("font_library_remove_app_action"));
        assertTrue(content.contains("dpisSegmentedShapes"));
        assertTrue(content.contains("FontFamily(typeface)"));
        assertFalse(detail.contains("createCompatibilitySection"));
        assertFalse(detail.contains("createManagementSection"));
        assertTrue(content.contains("ic_build_24"));
        assertTrue(content.contains("ic_edit_24"));
        assertTrue(content.contains("ic_delete_24"));
        assertTrue(content.contains("FontLibraryContentPreview"));
        assertTrue(content.contains("FontDetailContentPreview"));
        assertTrue(manifest.contains("android:name=\".fonts.FontDetailActivity\""));
        assertTrue(detail.contains("restoreTypefaceReferences(cleared)"));
        assertTrue(detail.contains("runtime state was not restored"));
        assertTrue(detail.contains("ComposeTextInputDialog.show(this"));
    }

    @Test
    public void deleteConfirmationsUseStandardDialogWidth() throws IOException {
        String source = read("src/main/java/com/dpis/module/fonts/FontDetailActivity.java");

        assertTrue(source.contains("private void confirmDelete("));
        assertTrue(source.contains("private void confirmForceDelete("));
        assertTrue(source.contains("showFontConfirmation(R.string.font_library_delete_title"));
        assertTrue(source.contains("ComposeConfirmDialog.showWithLabels"));
    }

    @Test
    public void renameAndRestoreDialogsUseStandardDialogWidth() throws IOException {
        String source = read("src/main/java/com/dpis/module/fonts/FontDetailActivity.java");

        assertTrue(source.contains("private void promptRename("));
        assertTrue(source.contains("private void confirmClearAppTypeface("));
        assertTrue(source.contains("ComposeTextInputDialog.show(this"));
        assertTrue(source.contains("R.string.font_library_restore_default_action"));
    }

    private static String read(String relativePath) throws IOException {
        return SourceSmokeTestPaths.read(relativePath);
    }

    private static int occurrences(String value, String needle) {
        int count = 0;
        int index = 0;
        while ((index = value.indexOf(needle, index)) >= 0) {
            count++;
            index += needle.length();
        }
        return count;
    }
}

File: app/src/test/java/com/dpis/module/fonts/FontLibraryActivitySourceSmokeTest.kt
package com.dpis.module

import org.junit.Assert.assertTrue
import org.junit.Test

class FontLibraryActivitySourceSmokeTest {
    @Test
    fun ttcImportRegistersAllLoadableFaces() {
        val source = read("src/main/java/com/dpis/module/fonts/FontLibraryActivity.kt")
        val importRules = read("src/main/java/com/dpis/module/fonts/FontImport.kt")
        val strings = read("src/main/res/values/strings.xml")

        source.assertContainsAll(
            "FontFileInspector.inspect(created)",
            "FontFileKind.TTC",
            "findLoadableTtcFaceIndexes(",
            "fontLibraryStore.registerCopiedFontFaces(",
            "FontTypefaceLoader.load(file, it)",
            "font_library_import_count_success",
        )
        importRules.assertContainsAll(
            "lowerName.endsWith(\".ttc\")",
            "mimeType == \"font/ttc\"",
        )
        source.assertNotContainsAll("isTtcFontImportEnabled", "showTtcFaceSelectionDialog")
        strings.assertNotContainsAll("font_library_ttc_select_title")
    }

    @Test
    fun fontPagesOwnDialogsInComposeState() {
        val library = read("src/main/java/com/dpis/module/fonts/FontLibraryActivity.kt")
        val detail = read("src/main/java/com/dpis/module/fonts/FontDetailActivity.kt")
        val content = read("src/main/java/com/dpis/module/fonts/presentation/FontLibraryContent.kt")
        val wear = read("src/main/java/com/dpis/module/ui/presentation/wear/WearSecondaryPages.kt")
        val textInput = read("src/main/java/com/dpis/module/ui/presentation/dialogs/TextInputDialog.kt")

        library.assertContainsAll(
            "FontLibraryDialog.Name(",
            "FontLibraryDialog.Large(",
            "FontLibraryDialog.Repair(",
        )
        detail.assertContainsAll(
            "FontDetailDialog.Rename(",
            "FontDetailDialog.Delete(",
            "FontDetailDialog.Restore(",
            "FontDetailDialog.Fallback",
        )
        content.assertContainsAll(
            "FontLibraryDialogHost(",
            "FontDetailDialogHost(",
            "TextInputDialog(",
            "ConfirmAlertDialog(",
        )
        wear.assertContainsAll("FontLibraryDialogHost(")
        textInput.assertContainsAll("fun TextInputDialog(", "ModalDialog(")
        library.assertNotContainsAll(
            "ComposeTextInputDialog",
            "ConfirmDialog.show",
            "MaterialAlertDialogBuilder",
        )
        detail.assertNotContainsAll(
            "ComposeTextInputDialog",
            "ConfirmDialog.show",
            "MaterialAlertDialogBuilder",
        )
        textInput.assertNotContainsAll("object ComposeTextInputDialog", "AlertDialog")
    }

    @Test
    fun oversizedFontUsesConfirmationInsteadOfHardRejection() {
        val source = read("src/main/java/com/dpis/module/fonts/FontLibraryActivity.kt")
        val importRules = read("src/main/java/com/dpis/module/fonts/FontImport.kt")
        val content = read("src/main/java/com/dpis/module/fonts/presentation/FontLibraryContent.kt")
        val strings = read("src/main/res/values/strings.xml")

        source.assertContainsAll(
            "FontImport.hasSpace(",
            "FontImport.LARGE_WARNING_BYTES",
            "confirmLargeFontImport(",
            "resolveDocumentSize(uri)",
        )
        importRules.assertContainsAll("sourceSizeBytes * 3L + FREE_SPACE_MARGIN_BYTES")
        content.assertContainsAll(
            "is FontLibraryDialog.Large",
            "font_library_large_import_message",
            "font_library_large_import_continue",
        )
        strings.assertContainsAll(
            "font_library_large_import_message",
            "font_library_import_insufficient_space",
        )
    }

    @Test
    fun fontDetailsDisplayPublicationStatus() {
        val source = read("src/main/java/com/dpis/module/fonts/FontDetailActivity.kt")
        val content = read("src/main/java/com/dpis/module/fonts/presentation/FontLibraryContent.kt")
        val strings = read("src/main/res/values/strings.xml")

        source.assertContainsAll(
            "FontDetailUiState(",
            "entry.sourceFileName.orEmpty()",
            "showFallbackExplanationDialog",
            "FontPublicationStatus.PUBLISH_FAILED",
            "retryPublishedFallbacks",
        )
        strings.assertContainsAll(
            "font_library_publication_retry_action",
            "font_library_fallback_dialog_message",
        )
        content.assertContainsAll(
            "state.publicationFailed",
            "font_library_used_badge",
        )
        content.assertNotContainsAll(
            "font_library_private_badge",
            "font_library_public_badge",
        )
        source.assertNotContainsAll("resolveFontSubtitle(entry)")
    }

    @Test
    fun fontLibraryGroupsFacesIntoCollections() {
        val source = read("src/main/java/com/dpis/module/fonts/FontLibraryActivity.kt")

        source.assertContainsAll(
            "linkedMapOf<String, MutableList<FontLibraryEntry>>()",
            "FontLibraryUiItem(",
            "faces.size > 1",
            "font_library_collection_label",
        )
    }

    @Test
    fun fontLibraryImportsAndExportsSeparateArchives() {
        val source = read("src/main/java/com/dpis/module/fonts/FontLibraryActivity.kt")
        val content = read("src/main/java/com/dpis/module/fonts/presentation/FontLibraryContent.kt")

        content.assertContainsAll(
            "ic_save_24",
            "font_library_export_archive_action",
            "font_library_import_archive_action",
        )
        source.assertContainsAll(
            "openFontLibraryExportPicker",
            "openFontLibraryImportPicker",
            "FontLibraryArchiveCodec.writeArchive",
            "FontLibraryArchiveCodec.restoreArchive",
            "REQUEST_EXPORT_FONT_LIBRARY",
            "REQUEST_IMPORT_FONT_LIBRARY",
            "recoverMissingFontCatalogAsync",
            "recoverMissingCatalogEntries",
        )
    }

    @Test
    fun fontDetailsUseDedicatedActivityAndFullPageLayout() {
        val library = read("src/main/java/com/dpis/module/fonts/FontLibraryActivity.kt")
        val detail = read("src/main/java/com/dpis/module/fonts/FontDetailActivity.kt")
        val content = read("src/main/java/com/dpis/module/fonts/presentation/FontLibraryContent.kt")
        val manifest = read("src/main/AndroidManifest.xml")

        library.assertContainsAll(
            "Intent(this, FontDetailActivity::class.java)",
            "FontDetailActivity.EXTRA_FONT_ID",
        )
        library.assertNotContainsAll("showFontDetails(")
        detail.assertContainsAll(
            "EXTRA_FONT_ID",
            "showFallbackExplanationDialog",
            "confirmDeleteForCurrentEntry",
            "confirmClearAppTypefaceByPackage",
            "restoreTypefaceReferences(cleared)",
            "runtime state was not restored",
        )
        content.assertContainsAll(
            "FontDetailCard(",
            "FontReferenceSection(",
            "font_library_used_by_title",
            "SegmentedListItem(",
            "dpisSegmentedShapes",
            "FontFamily(typeface)",
            "ToolbarIconButton(",
            "ic_edit_24",
            "ic_delete_24",
            "font_library_rename_action",
            "font_library_delete_action",
            "FontLibraryContentPreview",
            "FontDetailContentPreview",
        )
        content.assertNotContainsAll(
            "AssistChip(",
            "FontDetailHeader(",
            "BorderStroke",
            "DropdownMenu(",
            "font_library_detail_menu_action",
        )
        manifest.assertContainsAll("android:name=\".fonts.FontDetailActivity\"")
        detail.assertNotContainsAll("createCompatibilitySection", "createManagementSection")
    }

    @Test
    fun deleteAndRestoreConfirmationsUseComposeDialogState() {
        val source = read("src/main/java/com/dpis/module/fonts/FontDetailActivity.kt")
        val content = read("src/main/java/com/dpis/module/fonts/presentation/FontLibraryContent.kt")

        source.assertContainsAll(
            "confirmDeleteForCurrentEntry",
            "font_library_delete_title",
            "promptRename",
            "confirmClearAppTypefaceByPackage",
        )
        content.assertContainsAll(
            "is FontDetailDialog.Delete",
            "is FontDetailDialog.Restore",
            "is FontDetailDialog.Rename",
            "font_library_restore_default_action",
        )
        source.assertNotContainsAll("showFontConfirmation(", "confirmForceDelete(")
    }

    private fun read(relativePath: String) = SourceSmokeTestPaths.read(relativePath)

    private fun String.assertContainsAll(vararg needles: String) {
        needles.forEach { assertTrue("Missing $it", contains(it)) }
    }

    private fun String.assertNotContainsAll(vararg needles: String) {
        needles.forEach { assertTrue("Unexpected $it", !contains(it)) }
    }
}
